File: JavaSource/src/main/java/ru/akhcheck/patterns/mediator/dispatcher/PlaneDispatcher.java
package ru.akhcheck.patterns.mediator.dispatcher;

import java.util.ArrayList;
import java.util.List;
import ru.akhcheck.patterns.mediator.vehicles.Vehicle;

/**
 * Держит полосу и решает, кому её отдать.
 *
 * <p>Вся логика взаимодействия собрана здесь. Если бы самолёты договаривались
 * напрямую, каждому пришлось бы знать про остальных, и число связей росло бы
 * как квадрат числа участников.
 */
public class PlaneDispatcher implements Dispatcher {
    private final List<Vehicle> planes = new ArrayList<>();
    private Vehicle runwayTakenBy;

    public void addPlane(Vehicle plane) {
        planes.add(plane);
        plane.setDispatcher(this);
    }

    @Override
    public boolean notify(Vehicle vehicle, String message) {
        if ("request_landing".equals(message)) {
            if (runwayTakenBy == null) {
                runwayTakenBy = vehicle;
                System.out.println("  dispatcher: runway is given to " + vehicle.getName());
                return true;
            }
            System.out.println(
                    "  dispatcher: runway is taken by " + runwayTakenBy.getName()
                            + ", " + vehicle.getName() + " waits"
            );
            return false;
        }

        if ("runway_free".equals(message)) {
            if (runwayTakenBy == vehicle) {
                runwayTakenBy = null;
                System.out.println(
                        "  dispatcher: runway is free again after " + vehicle.getName()
                );
            }
            return true;
        }

        return false;
    }
}

File: Mediator/python-source/dispatcher/PlaneDispatcher.py
from __future__ import annotations

import sys
from pathlib import Path

if str(Path(__file__).resolve().parents[1]) not in sys.path:
    sys.path.append(str(Path(__file__).resolve().parents[1]))

from vehicles import Vehicle

from .Dispatcher import Dispatcher


class PlaneDispatcher(Dispatcher):
    """Держит полосу и решает, кому её отдать.

    Вся логика взаимодействия собрана здесь. Если бы самолёты договаривались
    напрямую, каждому пришлось бы знать про остальных, и число связей росло бы
    как квадрат числа участников.
    """

    def __init__(self) -> None:
        self._planes: list[Vehicle] = []
        self._runway_taken_by: Vehicle | None = None

    def add_plane(self, plane: Vehicle) -> None:
        self._planes.append(plane)
        plane.set_dispatcher(self)

    def notify(self, vehicle: Vehicle, message: str) -> bool:
        if message == "request_landing":
            if self._runway_taken_by is None:
                self._runway_taken_by = vehicle
                print(f"  dispatcher: runway is given to {vehicle.name}")
                return True
            print(
                f"  dispatcher: runway is taken by "
                f"{self._runway_taken_by.name}, {vehicle.name} waits"
            )
            return False

        if message == "runway_free":
            if self._runway_taken_by is vehicle:
                self._runway_taken_by = None
                print(f"  dispatcher: runway is free again after {vehicle.name}")
            return True

        return False
